src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,305 @@
+            : simdSet(simdRegistersNumber) {
+        checkUniqueAndUpdate();
+        for (auto& reg : regsToExclude) {
+            if (reg.isXMM() || reg.isYMM() || reg.isZMM()) {
+                simdSet.exclude(reg);
+            } else if (reg.isREG()) {
+                generalSet.exclude(reg);
+            }
+        }
+        generalSet.exclude(Xbyak::Reg64(Xbyak::Operand::RSP));
+    }
+
+private:
+    template<typename TReg>
+    int getFree(int requestedIdx) {
+        static_assert(std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value ||
+                      std::is_same<TReg, Xbyak::Reg8>::value || std::is_same<TReg, Xbyak::Reg16>::value ||
+                      std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Reg64>::value ||
+                      std::is_same<TReg, Xbyak::Opmask>::value,
+                      "Unsupported TReg by RegistersPool. Please, use the following Xbyak registers either "
+                      "Reg8, Reg16, Reg32, Reg64, Xmm, Ymm, Zmm or Opmask");
+        if (std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value) {
+            auto idx = simdSet.getUnused(requestedIdx);
+            simdSet.setAsUsed(idx);
+            return idx;
+        } else if (std::is_same<TReg, Xbyak::Reg8>::value || std::is_same<TReg, Xbyak::Reg16>::value ||
+                   std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Reg64>::value) {
+            auto idx = generalSet.getUnused(requestedIdx);
+            generalSet.setAsUsed(idx);
+            return idx;
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            return getFreeOpmask(requestedIdx);
+        }
+    }
+
+    template<typename TReg>
+    void returnToPool(TReg reg) {
+        if (std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value) {
+            simdSet.setAsUnused(reg.getIdx());
+        } else if (std::is_same<TReg, Xbyak::Reg8>::value || std::is_same<TReg, Xbyak::Reg16>::value ||
+                   std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Reg64>::value) {
+            generalSet.setAsUnused(reg.getIdx());
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            returnOpmaskToPool(reg.getIdx());
+        }
+    }
+
+    void checkUniqueAndUpdate(bool isCtor = true) {
+        static thread_local bool isCreated = false;
+        if (isCtor) {
+            if (isCreated) {
+                IE_THROW() << "There should be only one instance of RegistersPool per thread";
+            }
+            isCreated = true;
+        } else {
+            isCreated = false;
+        }
+    }
+
+    PhysicalSet generalSet {16};
+    PhysicalSet simdSet;
+};
+
+template <x64::cpu_isa_t isa>
+class IsaRegistersPool : public RegistersPool {
```

Comment:
Why not to make RegistersPool templated?

Reply:
Because in this version of C++ we do not have `if constexpr` and it would be hard to write code like this:
```cpp
if (mayiuse(avx512_core)) {
    regPool->countFree<Xbyak::OpMask>();
} else {
    regPool->countFree<Xbyak::Xmm>();
}
```

Reply:
I see 3 solutions for that:
1. Define smth like that:
```using Vmask = typename dnnl::impl::utils::conditional3<isa == dnnl::impl::cpu::x64::avx512_core, Xbyak::Opmask,
                                                           isa == dnnl::impl::cpu::x64::sse41,       Xbyak::Xmm,
                                                                                                     Xbyak::Ymm>::type;
```
2. Move this check into *countFree*:
``` ...
else if (std::is_same<TReg, Xbyak::Opmask>::value) {
            if (mayiuse(avx512_core)) {
                return countUnusedOpmask();
            } else {
                simdSet.countUnused();
            }
        }
```
3. Replace templated *countFree* by *countFreeRegs*, *countFreeSimds* and *countFreeMasks*. There is no different for class user how to call it: pass template parameter or write another name.
The 3rd one is more convenient for me. Actually countFree<Xbyak::Xmm>, countFree<Xbyak::Ymm>, countFree<Xbyak::Zmm> do the same as well as for general registers.  In worse case this function will be compiled for each register type that leads to meaningless binary size increasing.

Reply:
@nshchego  Just use
``` C++
regPool->countFree<Vmask>();
```
So, you could write generalized code.
for the current implementation of the RegisterPool.
I prepared the [PR](https://github.com/openvinotoolkit/openvino/pull/13070) with the RegisterPool usage examples.

Reply:
> @nshchego Just use
> 
> ```c++
> regPool->countFree<Vmask>();
> ```
> 
> So, you could write generalized code. for the current implementation of the RegisterPool. I prepared the [PR](https://github.com/openvinotoolkit/openvino/pull/13070) with the RegisterPool usage examples.

@nshchego For example
``` C++
if (regPool->countFree<Vmask>() > 0) {
    method_that_use_1_Vmask_register_from_pool();
} else if (regPool->countFree<Vmask>() == 0) {
    save_register_to_stack(someVmask);
    method_that_use_1_Vmask_register_from_pool();
    restore_register_from_stack(someVmask);
}
```
But with `countFreeSimds` and `countFreeMasks` such code will be much more complex.

Reply:
> I see 3 solutions for that:
> 
>     1. Define smth like that:
> 
> 
> ```
> using Vmask = typename dnnl::impl::utils::conditional3<isa == dnnl::impl::cpu::x64::avx512_core, Xbyak::Opmask,
>                                                            isa == dnnl::impl::cpu::x64::sse41,       Xbyak::Xmm,
>                                                                                                      Xbyak::Ymm>::type;
> ```
> 
>     2. Move this check into _countFree_:
> 
> 
> ```
> else if (std::is_same<TReg, Xbyak::Opmask>::value) {
>             if (mayiuse(avx512_core)) {
>                 return countUnusedOpmask();
>             } else {
>                 simdSet.countUnused();
>             }
>         }
> ```
> 
>     3. Replace templated _countFree_ by _countFreeRegs_, _countFreeSimds_ and _countFreeMasks_. There is no different for class user how to call it: pass template parameter or write another name.
> 
> 
> The 3rd one is more convenient for me. Actually countFreeXbyak::Xmm, countFreeXbyak::Ymm, countFreeXbyak::Zmm do the same as well as for general registers. In worse case this function will be compiled for each register type that leads to meaningless binary size increasing. _Binary size is one of the product criteria._

@nshchego I changed to use SFINAE, so only three countFree method instances will compile.

Reply:
ok, let's leave it